File: packages/axon-core-py/src/axon_core/station.py
from __future__ import annotations
from typing import Literal, Optional, Union
from pydantic import BaseModel, Field

CarrierType = Literal["tray", "pallet", "conveyor", "agv", "manual"]

TestKind = Literal[
    "visual", "dimensional", "leak-helium", "leak-pressure-decay",
    "isolation", "hipot", "hvil", "insulation-resistance",
    "weld-pull", "weld-resistance", "torque", "ocv", "dcir",
    "capacity", "thermal-imaging", "communication", "firmware-flash", "other",
]


class StationIO(BaseModel):
    itemRef: str
    carrier: CarrierType
    fixtureId: Optional[str] = None


class TestSpec(BaseModel):
    parameter: str
    unit: str
    min: Optional[float] = None
    max: Optional[float] = None
    rampMs: Optional[float] = None
    holdMs: Optional[float] = None
    samples: Optional[int] = None


class TestStep(BaseModel):
    id: str
    kind: TestKind
    spec: TestSpec
    required: bool = True
    runOnPriorFail: bool = False


class StationSpec(BaseModel):
    id: str
    name: str
    line: str
    position: int
    inputs: list[StationIO] = Field(default_factory=list)
    outputs: list[StationIO] = Field(default_factory=list)
    tests: list[TestStep] = Field(default_factory=list)
    cycleTimeSec: float
    takt: Optional[float] = None
    oprNum: Optional[int] = None
    site: Optional[str] = None


class TestResult(BaseModel):
    id: str
    stepId: str
    itemRef: str
    pass_: bool = Field(alias="pass")
    measurements: dict[str, Union[float, str]] = Field(default_factory=dict)
    startedAt: str
    finishedAt: str
    operator: Optional[str] = None
    notes: Optional[str] = None

    model_config = {"populate_by_name": True}

File: packages/axon-core-ts/src/station.ts
/** Station specification & test step model. Drives StationLine and EOL. */

export type CarrierType = "tray" | "pallet" | "conveyor" | "agv" | "manual";

export interface StationIO {
  itemRef: string;
  carrier: CarrierType;
  fixtureId?: string;
}

export type TestKind =
  | "visual"
  | "dimensional"
  | "leak-helium"
  | "leak-pressure-decay"
  | "isolation"
  | "hipot"
  | "hvil"
  | "insulation-resistance"
  | "weld-pull"
  | "weld-resistance"
  | "torque"
  | "ocv"
  | "dcir"
  | "capacity"
  | "thermal-imaging"
  | "communication"
  | "firmware-flash"
  | "other";

export interface TestSpec {
  parameter: string;
  unit: string;
  min?: number;
  max?: number;
  rampMs?: number;
  holdMs?: number;
  /** Number of samples / readings to acquire. */
  samples?: number;
}

export interface TestStep {
  id: string;
  kind: TestKind;
  spec: TestSpec;
  required: boolean;
  /** If a previous step failed, may we still attempt this one? */
  runOnPriorFail?: boolean;
}

export interface StationSpec {
  id: string;
  /** Human-readable name, e.g., "Cell Terminal Welding". */
  name: string;
  /** Logical line, e.g., "module-line-1". */
  line: string;
  /** Position in the line; lower runs first. */
  position: number;
  inputs: StationIO[];
  outputs: StationIO[];
  tests: TestStep[];
  cycleTimeSec: number;
  takt?: number; // target seconds per cycle for OEE tracking
  /** OprNum that this station maps to in the ERP routing. */
  oprNum?: number;
  /** Site code (e.g., factory or building). */
  site?: string;
}

export interface TestResult {
  id: string;
  stepId: string;
  itemRef: string;
  pass: boolean;
  measurements: Record<string, number | string>;
  startedAt: string;
  finishedAt: string;
  operator?: string;
  /** Free-form notes from operator or test runner. */
  notes?: string;
}
